libraries/rush-lib/src/cli/RushPnpmCommandLine.ts

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import {
+  FileSystem,
+  AlreadyReportedError,
+  Executable,
+  EnvironmentMap,
+  ITerminal,
+  Colors,
+  Terminal,
+  ConsoleTerminalProvider
+} from '@rushstack/node-core-library';
+import { PrintUtilities } from '@rushstack/terminal';
+
+import { RushConfiguration } from '../api/RushConfiguration';
+import { NodeJsCompatibility } from '../logic/NodeJsCompatibility';
+import { SpawnSyncReturns } from 'child_process';
+import { ILaunchOptions } from '../api/Rush';
+
+export interface ILaunchRushPnpmInternalOptions extends ILaunchOptions {}
+
+const RUSH_SKIP_CHECKS_PARAMETER: string = '--rush-skip-checks';
+
+enum CommandKind {
+  KnownSafe,
+  ShowWarning,
+  Blocked
+}
+
+export class RushPnpmCommandLine {
+  private static _knownCommandMap: Map<string, CommandKind> = new Map<string, CommandKind>([
+    ['add', CommandKind.Blocked],
+    ['audit', CommandKind.KnownSafe],
+    ['exec', CommandKind.KnownSafe],
+    ['import', CommandKind.Blocked],
+    ['install', CommandKind.Blocked],
+    [/* synonym */ 'i', CommandKind.Blocked],
+    ['install-test', CommandKind.Blocked],
+    [/* synonym */ 'it', CommandKind.Blocked],
+    ['link', CommandKind.ShowWarning],
+    [/* synonym */ 'ln', CommandKind.ShowWarning],
+    ['list', CommandKind.KnownSafe],
+    [/* synonym */ 'ls', CommandKind.KnownSafe],
+    ['outdated', CommandKind.KnownSafe],
+    ['pack', CommandKind.KnownSafe],
+    ['prune', CommandKind.KnownSafe],
+    ['publish', CommandKind.KnownSafe],
+    ['rebuild', CommandKind.KnownSafe],
+    [/* synonym */ 'rb', CommandKind.KnownSafe],
+    ['remove', CommandKind.ShowWarning],
+    [/* synonym */ 'rm', CommandKind.ShowWarning],
+    ['root', CommandKind.KnownSafe],
+    ['run', CommandKind.KnownSafe],
+    ['start', CommandKind.KnownSafe],
+    ['store', CommandKind.KnownSafe],
+    ['test', CommandKind.KnownSafe],
+    [/* synonym */ 't', CommandKind.KnownSafe],
+    ['unlink', CommandKind.ShowWarning],
+    ['update', CommandKind.ShowWarning],
+    [/* synonym */ 'up', CommandKind.ShowWarning],
+    ['why', CommandKind.KnownSafe]
+  ]);
+
+  public static launch(launcherVersion: string, options: ILaunchRushPnpmInternalOptions): void {
+    // Node.js can sometimes accidentally terminate with a zero exit code  (e.g. for an uncaught
+    // promise exception), so we start with the assumption that the exit code is 1
+    // and set it to 0 only on success.
+    process.exitCode = 1;
+
+    const { terminalProvider } = options;
+    const terminal: ITerminal = new Terminal(terminalProvider || new ConsoleTerminalProvider());
```

Comment:
`const { terminalProvider = new ConsoleTerminalProvider() } = options;` should perhaps be disallowed by our lint rules. 😄  

(I don't know what that syntax means, and if I learn, the knowledge could make me worse at judging whether code is easy to read heheh.)

Reply:
@octogonz See the section on "default values" https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
The syntax has been around for several years and is a standard part of destructuring. Importantly it allows `terminalProvider` to still be a `const` and now no longer have `undefined` as part of its type (since if it is `undefined`, the default expression evaluates during the assignment).

Reply:
(I chatted with @dmichon-msft about this and we agreed that `??` is a good compromise here.)

Reply:
Idea: `const { terminalProvider = new ConsoleTerminalProvider() } = options;` will create an instance of `ConsoleTerminalProvider` whether `terminalProvider` passed or not. So i use `||` instead. :)